tfjs-backend-webgpu/src/shader_preprocessor_wgsl.ts

```diff
@@ -0,0 +1,604 @@
+    [string, number] {
+  const {x, y = [], z = []} = dispatchLayout;
+
+  const outRank = outShape.length;
+  if (x.length === outRank) {
+    const dtype = getCoordsDataTypeWgsl(outRank);
+    const snippet = `fn getOutputCoords(globalId : vec3<u32>) -> ${dtype}{
+      return getCoordsFromFlatIndex(u32(globalId.x));
+    }
+    `;
+    return [snippet, outRank];
+  }
+
+  let gatherDimensionsStr = '';
+  const dims = [x, y, z];
+
+  let rank = 0;
+
+  for (let i = 0; i < dims.length; i++) {
+    const arr = dims[i];
+
+    if (arr.length === 0) {
+      continue;
+    }
+
+    rank += arr.length;
+
+    if (arr.length === 1) {
+      gatherDimensionsStr += `let d${arr[0]} : u32 =
+        u32(globalId[${i}]);`;
+    } else {
+      const strides = symbolicallyComputeStrides(arr, 'uniforms.outShape');
+      gatherDimensionsStr += `let index${i} : u32 =
+          u32(globalId[${i}]);`;
+      for (let j = 0; j < strides.length; j++) {
+        gatherDimensionsStr +=
+            `let d${arr[j]} : u32 = index${i} / ${strides[j]};`;
+
+        if (j === strides.length - 1) {
+          gatherDimensionsStr += `let d${arr[j + 1]} : u32 = ` +
+              `index${i} - d${arr[j]} * ${strides[j]};`;
+        } else {
+          gatherDimensionsStr +=
+              `index${i} = index${i} - d${arr[j]} * ${strides[j]};`;
+        }
+      }
+    }
+  }
+
+  const dimensions = [];
+  for (let i = 0; i < rank; i++) {
+    dimensions.push(`d${i}`);
+  }
+
+  const dtype = getCoordsDataTypeWgsl(rank);
+  let snippet = `fn getOutputCoords(globalId : vec3<u32>) -> ${dtype}{
+    ${gatherDimensionsStr}
+  `;
+  if (dimensions.length === 0) {
+    snippet += `return ${dtype}(0);}`;
+  } else {
+    snippet += `return ${dtype}(${dimensions.join(',')});}`;
+  }
+
+  return [snippet, rank];
+}
+
+/**
+ * Derives logical coordinates from a flat index. Performs integer division
+ * with each stride and decrements the index until the index equals the final
+ * dimension coordinate.
+ */
+function generateGetCoordsFromFlatIndex(shape: number[]): string {
+  const rank = shape.length;
+
+  if (rank <= 1) {
+    return `fn getCoordsFromFlatIndex(index : u32) ->u32 {return index; }`;
+  }
+
+  const strides = util.computeStrides(shape);
+  const dtype = getCoordsDataTypeWgsl(rank);
+
+  const coords: string[] = [];
+  for (let i = 0; i < rank; i++) {
+    coords.push(`d${i}`);
+  }
+
+  if (strides.length === 1) {
+    return `    fn getCoordsFromFlatIndex(index : u32) -> vec2<u32>{
+      let d0 = index / uniforms.outShapeStrides; let d1 = index - d0 * uniforms.outShapeStrides;
+      return vec2<u32>(d0,d1);
+    }`;
+  }
+  const snippet = 'var index2 : u32 = index;' +
+      strides
+          .map((_, i) => {
+            const line1 = `let ${
+                coords[i]} :u32 = index2 / uniforms.outShapeStrides[${i}]`;
+            const line2 = i === strides.length - 1 ?
+                `let ${coords[i + 1]}  : u32 = index2 - ${
```

Comment:
nit: Inconsistent spacing on `: u32`. You can just remove it, as the type should be inferred anyway.

Reply:
done